### PR TITLE
[feat]: Add support to external downloader aria2p

### PR DESCRIPTION
**Pull request by teddy171**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

This pull request support a external downloader aria2p(a command-line tool and Python library to interact with an aria2c daemon process through JSON-RPC). It can make download faster and get the rate of progress and other information such as download speed, eta time more easily(compared with use aria2c only). When use aria2p, install aria2c and aria2p(by pip) first. **Run aria2c and enable with the --enable-rpc option in the background.** Then all data will download with the aria2c daemon process.

For example:
```bash
python -m youtube_dl --external-downloader aria2p "http://www.youtube.com/watch?&v=BaW_jenozKc"
```
Related: #31188

### CODE REVIEW

**Review by dirkf — CHANGES_REQUESTED** [review]
Thanks for this interesting proposal.

Although (currently) it would only be useful with Python versions that are supported for _aria2fp_, I'd be happy to incorporate it as long as there are no unwanted side-effects when it's not available or supported.

Supposing someone has yt-dl installed with Python 2.7 but also has _aria2fp_ installed with a supported Python3, there could be an option to invoke _aria2fp_ by its CLI. I guess that would best be addressed with two FD classes, perhaps with some inheritance.

I've made some detailed suggestions to consider. Let me know what you think.

**Review by dirkf — COMMENTED** [review]
Somehow you've missed my existing review.

File: youtube_dl/downloader/external.py [hunk 1]
```diff
@@ -200,6 +205,74 @@ def _make_cmd(self, tmpfilename, info_dict):
...
+        if download_dir:
+            options["dir"] = download_dir
+        else:
+            options["dir"] = os.path.abspath('.')
+        options["out"] = os.path.basename(tmpfilename)
+        options["header"] = []
+        for key, val in info_dict['http_headers'].items():
+            options["header"].append(f"{key}: {val}")
+        download = aria2.add_uris([info_dict['url']], options)
+        while download.status == "active":
```
[the remark this answers is not in the document]
Comment (teddy171) [reply youtube_dl/downloader/external.py]: ```download``` do have a ```.status```. But its value could be ```waiting```, but when I test it, it seem return ```active``` immediately. Should I add ```waiting``` to the check?

**Comment by teddy171** [comment]
I think there is some problems. It raise ```AttributeError: 'Aria2pFD' object has no attribute '_filename'``` in line 256 external.py. Should I use the function ```undo_temp_name``` which in ```FileDownloader```?

File: youtube_dl/downloader/external.py [hunk 2]
```diff
@@ -200,6 +201,78 @@ def _make_cmd(self, tmpfilename, info_dict):
...
+        else:
+            options["dir"] = os.path.abspath('.')
+        options["out"] = os.path.basename(tmpfilename)
+        options["header"] = []
+        for key, val in info_dict['http_headers'].items():
+            options['header'].append('{0}: {1}'.format(key, val))
+        download = aria2.add_uris([info_dict['url']], options)
+        status = {
+            'status': 'downloading',
+            'filename': self._filename,
```
Comment (dirkf) [inline youtube_dl/downloader/external.py:258]:
```suggestion
            'filename': self.__filename,
```

File: youtube_dl/downloader/external.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (dirkf) [reply youtube_dl/downloader/external.py]: I think you should handle all the documented statuses that might be returned. I could see `waiting` in your animation.

If there's no documentation in the _aria2p_ project (or in the JSON-RPC spec for aria2), look under the hood of the _aria2p_ code to see what might appear.

I guess there will be `complete` that breaks the polling loop, plus some good values (`active`, `waiting`, ...?) that will continue the loop and anything else will be an error that breaks the loop.

File: youtube_dl/downloader/external.py (hunk 2 above)
Comment (dirkf) [reply youtube_dl/downloader/external.py]: That was the problem

**Comment by dirkf** [comment]
If you apply all the suggestions, and now that I've fixed the bad one, the CI tests might pass. Install _flake8_ with _pip_ and run _flake8_ on your code to identify code issues that will fail the Linter step (unused variables, indentation, etc).

You could make up a test harness like `youtube-dl/test/test_downloader_http.py` to check that your code actually works. The CI tests will only show that it doesn't break other things.

File: youtube_dl/downloader/external.py (hunk 2 above)
Comment (teddy171) [reply youtube_dl/downloader/external.py]:
```
AttributeError: 'Aria2pFD' object has no attribute '_Aria2pFD__filename'
```
Seem it still cannot work.

File: youtube_dl/downloader/external.py [hunk 3]
```diff
@@ -200,6 +200,69 @@ def _make_cmd(self, tmpfilename, info_dict):
...
+    (Aria2p, a command-line tool and Python library to interact with an aria2c daemon process
+    through JSON-RPC.)
+    It can help you to get download progress more easily.
+    To use aria2p as downloader, you need to install aria2c and aria2p, aria2p can download with pip.
+    Then run aria2c in the background and enable with the --enable-rpc option.
+    '''
+    try:
+        import aria2p
+        __avail = True
+    except ModuleNotFoundError:
```
Comment (dirkf) [inline youtube_dl/downloader/external.py:215]:
```suggestion
    except ImportError:
```

**Comment by rautamiekka** [comment]
Surely using PyCharm, being a very powerful dedicated Python IDE, would help a lot.

**Comment by dirkf** [comment]
A test harness can be committed into the repository and used to test regressions (though in this case it would have to be disabled from the automatic CI tests).

**Comment by teddy171** [comment]
Is it OK to merge?

**Comment by dirkf** [comment]
Make your test succeed trivially if the aria2p package isn't available, but run fully if it is. Post the successful test log from your environment.

**Comment by teddy171** [comment]
My test log
```bash
$ python -m unittest test/test_external_downloader.py 
..
----------------------------------------------------------------------
Ran 2 tests in 0.667s

OK
```
Is it OK?

File: test/test_external_downloader.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+                    params['logger'] = FakeLogger()
+                    params['outtmpl'] = filename
+                    ydl = YoutubeDL(params)
+                    try_rm(encodeFilename(filename))
+                    self.assertEqual(ydl.download(['http://127.0.0.1:%d/%s' % (self.port, ep)]), 0)
+                    self.assertEqual(os.path.getsize(encodeFilename(filename)), TEST_SIZE)
+                    try_rm(encodeFilename(filename))
+                process.kill()
+        except Exception:
+            pass
```
Comment (dirkf) [inline test/test_external_downloader.py:106]: Don't we want know if this failed?
```suggestion
```

File: test/test_external_downloader.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+class HTTPTestRequestHandler(compat_http_server.BaseHTTPRequestHandler):
+    def log_message(self, format, *args):
+        pass
+
+    def send_content_range(self, total=None):
+        range_header = self.headers.get('Range')
+        start = end = None
+        if range_header:
+            mobj = re.search(r'^bytes=(\d+)-(\d+)', range_header)
```
Comment (dirkf) [inline test/test_external_downloader.py:32]:
```suggestion
            mobj = re.match(r'bytes=(\d+)-(\d+)', range_header)
```

File: test/test_external_downloader.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+            if mobj:
+                start = int(mobj.group(1))
+                end = int(mobj.group(2))
```
Comment (dirkf) [inline test/test_external_downloader.py:35]:
```suggestion
                start, end = (int(mobj.group(i)) for i in (1, 2))
```

File: test/test_external_downloader.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        if self.path == '/regular':
+            self.serve()
+        elif self.path == '/no-content-length':
+            self.serve(content_length=False)
+        elif self.path == '/no-range':
+            self.serve(range=False)
+        elif self.path == '/no-range-no-content-length':
+            self.serve(range=False, content_length=False)
+        else:
+            assert False
```
Comment (dirkf) [inline test/test_external_downloader.py:65]:
```suggestion
            assert False, 'unrecognised server path'
```

File: test/test_external_downloader.py [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+        pass
+
+    def warning(self, msg):
+        pass
+
+    def error(self, msg):
+        pass
+
+
+class TestAria2pFD(unittest.TestCase):
```
Comment (dirkf) [inline test/test_external_downloader.py:79]: Automate skipping, which will document whether the test ran fully:
```suggestion
@unittest.skipUnless(Aria2pFD.available(), 'aria2p module not found')
class TestAria2pFD(unittest.TestCase):
```

File: youtube_dl/downloader/external.py [hunk 9]
```diff
@@ -200,6 +200,69 @@ def _make_cmd(self, tmpfilename, info_dict):
...
+    except ImportError:
+        __avail = False
+
+    @classmethod
+    def available(cls):
+        return cls.__avail
+
+    def _call_downloader(self, tmpfilename, info_dict):
+        import aria2p
+        aria2 = aria2p.API(
```
Comment (dirkf) [inline youtube_dl/downloader/external.py:224]: The `aria2p` module namespace was already imported:
```suggestion
        aria2 = self.aria2p.API(
```

File: youtube_dl/downloader/external.py [hunk 10]
```diff
@@ -200,6 +200,69 @@ def _make_cmd(self, tmpfilename, info_dict):
...
+            'max-connection-per-server': 4,
+            'auto-file-renaming': 'false',
+        }
+        download_dir = os.path.dirname(tmpfilename)
+        if download_dir:
+            options["dir"] = download_dir
+        else:
+            options["dir"] = os.path.abspath('.')
+        options["out"] = os.path.basename(tmpfilename)
+        options["header"] = []
```
Comment (dirkf) [inline youtube_dl/downloader/external.py:243]: Use `'`: project convention.

**Review by dirkf — CHANGES_REQUESTED** [review]
Almost ready to go, I hope.

**Comment by teddy171** [comment]
Thank you for your suggestions @dirkf. I have applied your suggestions. Is it able to merge now?

File: test/helper.py [hunk 11]
```diff
@@ -88,6 +88,15 @@ def report_warning(self, message):
             old_report_warning(message)
         self.report_warning = types.MethodType(report_warning, self)
```
Comment (dirkf) [inline test/helper.py:90]:
```suggestion


```

File: test/helper.py [hunk 12]
```diff
@@ -88,6 +88,15 @@ def report_warning(self, message):
...
 
+class FakeLogger(object):
+    def debug(self, msg):
+        pass
+
+    def warning(self, msg):
+        pass
+
+    def error(self, msg):
+        pass
```
Comment (dirkf) [inline test/helper.py:100]:
```suggestion


```

File: test/test_downloader_http.py [hunk 13]
```diff
@@ -9,7 +9,7 @@
 import unittest
 sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
 
-from test.helper import http_server_port, try_rm
+from test.helper import FakeLogger, http_server_port, try_rm
```
Comment (dirkf) [inline test/test_downloader_http.py:12]:
```suggestion
from test.helper import (
    FakeLogger,
    http_server_port,
    try_rm,
)
```

File: test/test_external_downloader.py [hunk 14]
```diff
@@ -0,0 +1,110 @@
...
+
+# Allow direct execution
+import os
+import re
+import sys
+import subprocess
+import unittest
+sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+
+from test.helper import FakeLogger, http_server_port, try_rm
```
Comment (dirkf) [inline test/test_external_downloader.py:13]:
```suggestion
from test.helper import (
    FakeLogger,
    http_server_port,
    try_rm,
)
```

File: test/test_http.py [hunk 15]
```diff
@@ -8,7 +8,7 @@
 import unittest
 sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
 
-from test.helper import http_server_port
+from test.helper import http_server_port, FakeLogger
```
Comment (dirkf) [inline test/test_http.py:11]:
```suggestion
from test.helper import (
    FakeLogger,
    http_server_port,
)
```

File: youtube_dl/downloader/external.py [hunk 16]
```diff
@@ -200,6 +200,68 @@ def _make_cmd(self, tmpfilename, info_dict):
...
+        options = {
+            'min-split-size': '1M',
+            'max-connection-per-server': 4,
+            'auto-file-renaming': 'false',
+        }
+        download_dir = os.path.dirname(tmpfilename)
+        if download_dir:
+            options['dir'] = download_dir
+        else:
+            options['dir'] = os.path.abspath('.')
```
Comment (dirkf) [inline youtube_dl/downloader/external.py:240]:
```suggestion
        options['dir'] = os.path.dirname(tmpfilename) or os.path.abspath('.')
```

File: youtube_dl/downloader/external.py [hunk 17]
```diff
@@ -200,6 +200,68 @@ def _make_cmd(self, tmpfilename, info_dict):
...
+            status.update({
+                'downloaded_bytes': download.completed_length,
+                'total_bytes': download.total_length,
+                'elapsed': time.time() - started,
+                'eta': download.eta.total_seconds(),
+                'speed': download.download_speed,
+            })
+            self._hook_progress(status)
+            time.sleep(.5)
+        return not download.status == 'complete'
```
Comment (dirkf) [inline youtube_dl/downloader/external.py:262]:
```suggestion
        return download.status != 'complete'
```

File: test/test_external_downloader.py [hunk 18]
```diff
@@ -0,0 +1,110 @@
...
+from youtube_dl.compat import compat_http_server
+from youtube_dl.utils import encodeFilename
+from youtube_dl.downloader.external import Aria2pFD
+import threading
+
+TEST_DIR = os.path.dirname(os.path.abspath(__file__))
+
+
+TEST_SIZE = 10 * 1024
+
```
Comment (dirkf) [inline test/test_external_downloader.py:24]:
```suggestion


```

**Review by dirkf — COMMENTED** [review]
I'll just apply some Lint, style and simplification fixes. If the CI tests work, please post your test output too. If all good we can merge.

**Comment by teddy171** [comment]
The new code is OK. Thank you @dirkf. Test Log
```bash
$ python -m unittest test/test_external_downloader.py 
..
----------------------------------------------------------------------
Ran 2 tests in 0.586s

OK
```

**Review by dirkf — APPROVED** [review]

**Comment by dirkf** [comment]
And in the CI tests, as expected:
```
test_chunked (test.test_external_downloader.TestAria2pFD) ... SKIP: aria2p module not found
test_regular (test.test_external_downloader.TestAria2pFD) ... SKIP: aria2p module not found
```
On reflection the test file ought to be named like the one for HTTP, so I did that.

**Pull request merged**